File: python/scripts/prove_prod.py
"""prove_prod — End-to-end smoke proof for the Python SDK against live prod.

What it does:
    1. Imports the public SDK surface (Heroa, DeployRequest, FileOverlay).
    2. Reads HEROA_API_KEY + HEROA_BASE_URL + HEROA_REGION from env.
    3. Calls Heroa.deploy(DeployRequest(template="static-site", files=[...])).
    4. Polls GET /v1/apps/{app_id}/machines/{id} until observed_state ==
       running, or the highest reached state in {created, starting, running}.
    5. Cleans up: Heroa.stop, then DELETE /v1/apps/{app_id}.
    6. Writes proof lines to stdout via sys.stdout.write (the caller
       redirects to audit/sdk-python-prove.md).

Run:  HEROA_API_KEY=... python -m heroa_sdk_prove_prod
or:   HEROA_API_KEY=... python sdk/python/scripts/prove_prod.py
"""

from __future__ import annotations

import json
import os
import sys
import time

import httpx

from heroa_sdk import DeployRequest, FileOverlay, Heroa

PROD_CP = "https://heroa-cp-2sobff3gmq-uc.a.run.app"
APP_PREFIX = "py-sdk-prove"
DEFAULT_REGION = "us-central1"
POLL_TIMEOUT_S = 60.0
POLL_INTERVAL_S = 2.0


def emit(line: str) -> None:
    sys.stdout.write(line + "\n")
    sys.stdout.flush()


def lookup_app_id(base_url: str, api_key: str, app_name: str) -> str:
    resp = httpx.post(
        f"{base_url}/v1/apps",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        content=json.dumps({"app_name": app_name, "org_slug": ""}).encode("utf-8"),
        timeout=30.0,
    )
    if resp.status_code not in (200, 201):
        raise RuntimeError(f"app lookup HTTP {resp.status_code}: {resp.text}")
    body = resp.json()
    app_id = body.get("id")
    if not app_id:
        raise RuntimeError(f"app id missing in response: {body!r}")
    return str(app_id)


def get_machine(base_url: str, api_key: str, app_id: str, machine_id: str) -> dict:
    resp = httpx.get(
        f"{base_url}/v1/apps/{app_id}/machines/{machine_id}",
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=30.0,
    )
    if resp.status_code != 200:
        raise RuntimeError(f"poll machine HTTP {resp.status_code}: {resp.text}")
    return dict(resp.json())


def delete_app(base_url: str, api_key: str, app_id: str) -> tuple[int, str]:
    resp = httpx.delete(
        f"{base_url}/v1/apps/{app_id}",
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=30.0,
    )
    return resp.status_code, resp.text.strip()


def main() -> int:
    api_key = os.environ.get("HEROA_API_KEY")
    if not api_key:
        emit("FAIL: HEROA_API_KEY is required (mint via POST /internal/api-keys)")
        return 1
    base_url = os.environ.get("HEROA_BASE_URL", PROD_CP)
    region = os.environ.get("HEROA_REGION", DEFAULT_REGION)

    ts = int(time.time())
    marker = f"PY-SDK-PROVE-{ts}"
    app_name = f"{APP_PREFIX}-{ts}"
    index_html = (
        "<!DOCTYPE html>\n"
        f"<html><head><title>{marker}</title></head>"
        f"<body><h1>{marker}</h1></body></html>\n"
    )

    emit("# Python SDK prove-prod")
    emit(f"base_url:  {base_url}")
    emit(f"app_name:  {app_name}")
    emit(f"region:    {region}")
    emit(f"marker:    {marker}")
    emit("sdk_path:  heroa_sdk (Heroa.deploy)")
    emit("")

    with Heroa(api_key, base_url=base_url, default_app_name=app_name) as client:
        start_call = time.time()
        instance = client.deploy(
            DeployRequest(
                template="static-site",
                region=region,
                app_name=app_name,
                size="small",
                files=[FileOverlay(path="/srv/index.html", content=index_html)],
                metadata={"prove": "py-sdk", "marker": marker},
                env={"HEROA_PROVE_MARKER": marker},
                restart_policy="always",
            )
        )
        deploy_ms = int((time.time() - start_call) * 1000)

        emit("## deploy() result")
        emit(f"machine_id:        {instance.id}")
        emit(f"region:            {instance.region}")
        emit(f"hostnames:         {list(instance.hostnames)}")
        emit(f"url:               {instance.url}")
        emit(f"size:              {instance.size}")
        emit(f"state (initial):   {instance.state}")
        emit(f"created_at:        {instance.created_at}")
        emit(f"deploy_call_ms:    {deploy_ms}")
        emit("")

        if not instance.id:
            emit("FAIL: deploy returned empty machine id")
            return 1

        app_id = lookup_app_id(base_url, api_key, app_name)

        deadline = time.time() + POLL_TIMEOUT_S
        last: dict = {}
        attempts = 0
        reached = "other"
        while time.time() < deadline:
            attempts += 1
            last = get_machine(base_url, api_key, app_id, instance.id)
            obs = last.get("observed_state", "")
            if obs == "running":
                reached = "running"
                break
            if obs == "created" and reached not in ("starting", "running"):
                reached = "created"
            elif obs == "starting" and reached != "running":
                reached = "starting"
            time.sleep(POLL_INTERVAL_S)

        emit("## poll")
        emit(f"attempts:        {attempts}")
        emit(f"final state:     {last.get('state', '')}")
        emit(f"observed_state:  {last.get('observed_state', '')}")
        emit(f"desired_state:   {last.get('desired_state', '')}")
        emit(f"reached:         {reached}")
        emit("")

        if reached == "other":
            emit(f"FAIL: machine never reached running/created/starting (last={last!r})")
            return 1

        emit("## cleanup")
        client.stop(app_name, instance.id)
        emit(f"stop({app_name}, {instance.id}): OK")

        del_status, del_body = delete_app(base_url, api_key, app_id)
        emit(f"DELETE /v1/apps/{app_id}: HTTP {del_status}")
        emit(f"  body: {del_body}")

    emit("")
    emit("## proof")
    emit("SDK:             heroa_sdk (Python)")
    emit(f"marker:          {marker}")
    emit(f"machine_id:      {instance.id}")
    emit("tenant:          (resolved by bearer; spec-13 e2e tenant)")
    emit("result:          PASS — deploy/poll/cleanup round-trip OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: typescript/scripts/prove-prod.ts
// prove-prod.ts — End-to-end smoke proof against the live prod control plane.
//
// What it does:
//   1. Imports the public TS SDK surface (Heroa client + deploy primitive).
//   2. Reads HEROA_API_KEY + HEROA_BASE_URL from env (or falls back to
//      the prod CP). The bearer must be minted out-of-band via
//      `POST /internal/api-keys` against the test tenant.
//   3. Calls heroa.deploy({ template: 'static-site', files: [...] }) with a
//      unique marker string that round-trips into /srv/index.html.
//   4. Polls GET /v1/apps/{app}/machines/{id} until observed_state==running
//      or a deadline elapses; asserts at least observed_state in
//      {created,starting,running}.
//   5. Cleans up: DELETE machine + DELETE app.
//   6. Writes proof lines to stdout (caller redirects to audit/sdk-typescript-prove.md).
//
// Run:  HEROA_API_KEY=... npx tsx scripts/prove-prod.ts

import { Heroa, type Region } from "../src/index.js";

const PROD_CP = "https://heroa-cp-2sobff3gmq-uc.a.run.app";
const APP_PREFIX = "ts-sdk-prove";
// us-central1 is where the substrate fleet currently has Firecracker
// capacity. us-east + us-west are tenant-allowlisted but the placement
// fleet returns region_capacity until those regions are stood up.
const REGION_DEFAULT = "us-central1";
const REGION = process.env.HEROA_REGION ?? REGION_DEFAULT;
const POLL_TIMEOUT_MS = 60_000;
const POLL_INTERVAL_MS = 2_000;

interface MachineWire {
  id: string;
  state: string;
  observed_state: string;
  desired_state: string;
}

function emit(line: string): void {
  process.stdout.write(line + "\n");
}

async function main(): Promise<void> {
  const apiKey = process.env.HEROA_API_KEY;
  if (!apiKey) {
    throw new Error("HEROA_API_KEY is required (mint via POST /internal/api-keys)");
  }
  const baseUrl = process.env.HEROA_BASE_URL ?? PROD_CP;

  const ts = Math.floor(Date.now() / 1000);
  const marker = `TS-SDK-PROVE-${ts}`;
  const appName = `${APP_PREFIX}-${ts}`;
  const indexHtml =
    `<!DOCTYPE html>\n<html><head><title>${marker}</title></head>` +
    `<body><h1>${marker}</h1></body></html>\n`;

  const heroa = new Heroa({
    apiKey,
    baseUrl,
    defaultAppName: appName,
  });

  emit(`# TS SDK prove-prod`);
  emit(`base_url:  ${baseUrl}`);
  emit(`app_name:  ${appName}`);
  emit(`region:    ${REGION}`);
  emit(`marker:    ${marker}`);
  emit(`sdk_path:  @heroa/sdk (Heroa, deploy)`);
  emit("");

  const startCall = Date.now();
  const instance = await heroa.deploy({
    template: "static-site",
    region: REGION as Region,
    appName,
    size: "small",
    files: [{ path: "/srv/index.html", content: indexHtml }],
    metadata: { "prove": "ts-sdk", "marker": marker },
    env: { "HEROA_PROVE_MARKER": marker },
    restartPolicy: "always",
  });
  const deployMs = Date.now() - startCall;

  emit(`## deploy() result`);
  emit(`machine_id:        ${instance.id}`);
  emit(`region:            ${instance.region}`);
  emit(`hostnames:         ${JSON.stringify(instance.hostnames)}`);
  emit(`url:               ${instance.url}`);
  emit(`size:              ${instance.size}`);
  emit(`state (initial):   ${instance.state}`);
  emit(`created_at:        ${instance.createdAt}`);
  emit(`deploy_call_ms:    ${deployMs}`);
  emit("");

  if (!instance.id) {
    throw new Error("deploy returned no machine id");
  }

  // The CP routes /v1/apps/{app}/... by the app id, not the name.
  // Look up the app's id via the idempotent createApp endpoint so the
  // poll + delete-app calls below address the right resource.
  const appLookup = await fetch(`${baseUrl}/v1/apps`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ app_name: appName, org_slug: "" }),
  });
  if (appLookup.status !== 200 && appLookup.status !== 201) {
    throw new Error(`app lookup: HTTP ${appLookup.status} ${await appLookup.text()}`);
  }
  const appBody = (await appLookup.json()) as { id: string };
  const appId = appBody.id;

  // Poll the machine state via the same bearer.
  const machineUrl =
    `${baseUrl}/v1/apps/${encodeURIComponent(appId)}/machines/${encodeURIComponent(instance.id)}`;
  const deadline = Date.now() + POLL_TIMEOUT_MS;
  let last: MachineWire | null = null;
  let pollAttempts = 0;
  let reached: "running" | "created" | "starting" | "other" = "other";
  while (Date.now() < deadline) {
    pollAttempts += 1;
    const res = await fetch(machineUrl, {
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    if (res.status !== 200) {
      throw new Error(`poll machine: HTTP ${res.status} ${await res.text()}`);
    }
    last = (await res.json()) as MachineWire;
    if (last.observed_state === "running") {
      reached = "running";
      break;
    }
    if (last.observed_state === "created") {
      reached = "created";
    } else if (last.observed_state === "starting") {
      reached = "starting";
    }
    await new Promise((r) => setTimeout(r, POLL_INTERVAL_MS));
  }

  emit(`## poll`);
  emit(`attempts:        ${pollAttempts}`);
  emit(`final state:     ${last?.state}`);
  emit(`observed_state:  ${last?.observed_state}`);
  emit(`desired_state:   ${last?.desired_state}`);
  emit(`reached:         ${reached}`);
  emit("");

  if (reached === "other") {
    throw new Error(
      `machine never reached running/created/starting (last=${JSON.stringify(last)})`,
    );
  }

  // Cleanup: delete machine, then delete app.
  emit(`## cleanup`);
  await heroa.stop(appName, instance.id);
  emit(`stop(${appName}, ${instance.id}): OK`);

  const delAppRes = await fetch(`${baseUrl}/v1/apps/${encodeURIComponent(appId)}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  emit(`DELETE /v1/apps/${appId}: HTTP ${delAppRes.status}`);
  if (delAppRes.status !== 200 && delAppRes.status !== 204) {
    emit(`  body: ${await delAppRes.text()}`);
  }

  emit("");
  emit(`## proof`);
  emit(`SDK:             @heroa/sdk (TypeScript)`);
  emit(`marker:          ${marker}`);
  emit(`machine_id:      ${instance.id}`);
  emit(`tenant:          (resolved by bearer; spec-13 e2e tenant)`);
  emit(`result:          PASS — deploy/poll/cleanup round-trip OK`);
}

main().catch((err) => {
  process.stderr.write(
    "FAIL: " + (err && (err as Error).stack ? (err as Error).stack : String(err)) + "\n",
  );
  process.exit(1);
});
